include/CObjectCache.h

```diff
@@ -9,58 +9,110 @@
         ContainerT m_container;
 
-        using ValueType = typename ContainerT_T<K..., T>::value_type::second_type; // container's value_type is always instantiation of std::pair
-        using NoPtrValueType = typename std::remove_pointer<ValueType>::type;
-        using PtrToConstVal_t = const NoPtrValueType*; // ValueType is always pointer type
+        // typedefs for implementation only
+        using PairType_impl = typename ContainerT::value_type;
+        //! Always pointer type
+        using ValueType_impl = typename PairType_impl::second_type; // container's value_type is always instantiation of std::pair
+        static_assert(std::is_pointer<ValueType_impl>::value, "ValueType_impl must be pointer type!");
+        using KeyType_impl = typename PairType_impl::first_type;
+        using NoPtrValueType_impl = typename std::remove_pointer<ValueType_impl>::type;
+        using ValueType_PtrToConst_impl = const NoPtrValueType_impl*; // ValueType_impl is always pointer type
+
+    public:
+        using RangeType = std::pair<IteratorType, IteratorType>;
+        using ConstRangeType = std::pair<ConstIteratorType, ConstIteratorType>;
+
+        using GreetFuncType = std::function<void(ValueType_impl)>;
+        using DisposalFuncType = std::function<void(ValueType_impl)>;
+
+        template<typename RangeT>
+        static bool isNonZeroRange(const RangeT& _range) { return std::distance(_range.first, _range.second) != 0; }
 
-        std::function<void(ValueType)> m_greetingFunc, m_disposalFunc;
+    protected:
+        GreetFuncType m_greetingFunc;
+        DisposalFuncType m_disposalFunc;
     protected:
 		inline virtual ~CObjectCacheBase()
 		{
 			for (auto it=m_container.begin(); it!=m_container.end(); it++)
 				dispose(it->second);
 		}
 
-        void dispose(ValueType _object) const        {
+        void dispose(ValueType_impl _object) const        {
             if (m_disposalFunc)
                 m_disposalFunc(_object);
         }
 
-        void greet(ValueType _object) const
+        void greet(ValueType_impl _object) const
         {
             if (m_greetingFunc)
                 m_greetingFunc(_object);
         }
 
     public:
-        inline explicit CObjectCacheBase(const std::function<void(ValueType)>& _greeting, const std::function<void(ValueType)>& _disposal) : m_greetingFunc(_greeting), m_disposalFunc(_disposal) {}
-        inline explicit CObjectCacheBase(std::function<void(ValueType)>&& _greeting, std::function<void(ValueType)>&& _disposal) : m_greetingFunc(std::move(_greeting)), m_disposalFunc(std::move(_disposal)) {}
+        inline explicit CObjectCacheBase(const GreetFuncType& _greeting, const DisposalFuncType& _disposal) : m_greetingFunc(_greeting), m_disposalFunc(_disposal) {}
+        inline explicit CObjectCacheBase(GreetFuncType&& _greeting, DisposalFuncType&& _disposal) : m_greetingFunc(std::move(_greeting)), m_disposalFunc(std::move(_disposal)) {}
```

Comment:
a move constructor would require you to zero out or invalidate the input reference.

Also why not copy and move constructors for the cache base itself?

Reply:
>a move constructor would require you to zero out or invalidate the input reference.

That's why I use `std::move`. So `std::function` takes care of that

>Also why not copy and move constructors for the cache base itself?

Sorry, I don't understand what you're asking for